[ADD] event_registration_cancel_reason

pedrobaeza opened the pull request:
# 
# Reasons for event registrations cancellations

This module allows to add a predefined reason why a registration for an event
is cancelled.
# Configuration

Go to _Marketing > Configuration > Events > Registrations cancel reasons_ for
setting the possible reasons you want to handle. You can also set if a reason
is only available for certain events or, if you leave it empty, for all.
# Usage

Go to _Marketing > Events > Events_, enter into the details or any of the
events, and press the red circle button in the right part of any of the
registrations. A new screen will appear asking you to enter the reason why the
registration is cancelled.

You will get also the same screen when pressing the button
_Cancel registration_ from the details view of a registration in
_Marketing > Events > Registrations_. On this object, you can group by the
cancel reason in the list, calendar or graph view.

pedrobaeza committed:
[ADD] event_registration_cancel_reason

=============================================
Reasons for event registrations cancellations
=============================================

This module allows to add a predefined reason why a registration for an event
is cancelled.

Configuration
=============

Go to *Marketing > Configuration > Events > Registrations cancel reasons* for
setting the possible reasons you want to handle. You can also set if a reason
is only available for certain events or, if you leave it empty, for all.

Usage
=====

Go to *Marketing > Events > Events*, enter into the details or any of the
events, and press the red circle button in the right part of any of the
registrations. A new screen will appear asking you to enter the reason why the
registration is cancelled.

You will get also the same screen when pressing the button
*Cancel registration* from the details view of a registration in
*Marketing > Events > Registrations*. On this object, you can group by the
cancel reason in the list, calendar or graph view.

pedrobaeza commented:
cc @rafaelbn

rafaelbn commented:
Don't you think is better to have reason per type of event an not per event?

![03-02-2016 10-47-01](https://cloud.githubusercontent.com/assets/8736623/12778586/b4cbadda-ca63-11e5-831b-a38f494fadf1.png)

pedrobaeza commented:
Yes, I haven't thought about it...

rafaelbn commented:
It could be very nice to have this in Event analysis, where a user will do full reporting

![03-02-2016 10-51-25](https://cloud.githubusercontent.com/assets/8736623/12778681/5c55cd42-ca64-11e5-88cc-27439c5ded97.png)

pedrobaeza commented:
It can't, as this analysis is for events, not for registrations. You already have it in the graph view of registrations.

rafaelbn commented:
OK. Functionally tested :+1: 
Just change commented in https://github.com/OCA/event/pull/30#issuecomment-179131975
Thanks!

pedrobaeza committed:
[IMP] event_registration_cancel_reason: Use event types instead of events

pedrobaeza commented:
Changed constraint in reasons to event types.

andhit-r commented:
@pedrobaeza when registration is cancelled I don't see the cancellation reason on form. IMHO, It should show somewhere on the form and/or tree, like Opportunity/Sale cancel reason.

pedrobaeza commented:
It's already in the form view of the registration. Do you mean any other side?

rafaelbn commented:
what @andhit-r wants to say it that after last commit this fields has disapeared

![03-02-2016 14-44-43](https://cloud.githubusercontent.com/assets/8736623/12783980/c7b90d04-ca84-11e5-80fb-81ebd96cd33a.png)

andhit-r commented:
@rafaelbn I checked again, now it shows. Strange. I'm quite sure i didn't see the cancel reason before.

andhit-r commented:
@pedrobaeza @rafaelbn I know what happend. It seems that field only can be seen by user on technical feauture group. By default admin doesn't belong to that group.

andhit-r commented:
:+1:

JavierIniesta commented:
Tested in runbot :+1:

JMODaERP commented:
I´ve tested and O.K.  I have created a new reason and I´ve canceled from Event and from Registration. :+1:

rafaelbn commented:
OK, then as said in https://github.com/OCA/event/pull/30#issuecomment-179250541 we must wait, because that field must be seen by a Marketing -> User

yajo commented on event_registration_cancel_reason/models/event_registration.py:
```diff
@@ -0,0 +1,37 @@
+    _inherit = 'event.registration'
+
+    cancel_reason_id = fields.Many2one(
+        comodel_name='event.registration.cancel.reason', readonly=True,
```
IMHO readonly and visibility should depend on the state.

pedrobaeza replied:
No, because it can only be set on the wizard. Or at least this is how it has been thought

yajo replied:
Well actually with visibility should be enough if you remove readonly.

yajo replied:
What if you made a mistake?

pedrobaeza replied:
Go back the registration to draft, and then cancel again. I think it's better than changing directly the field, but I don't have any strong feeling about that. It's more a user flow/opinion.

yajo commented on event_registration_cancel_reason/views/event_registration_view.xml:
```diff
@@ -0,0 +1,51 @@
+        <field name="inherit_id" ref="event.view_event_registration_form"/>
+        <field name="arch" type="xml">
+            <field name="date_closed" position="after">
+                <field name="cancel_reason_id"/>
```
I cannot see this in runbot.

pedrobaeza replied:
It seems that it's masked by "Technical features" security group for the position chosen

yajo replied:
That does not seem a technical feature, could you put it in another place?

pedrobaeza replied:
Yes, but maybe you can make a PR over my branch to complete comments as I'm now doing another things.

andhit-r committed:
1. Move cancel_reason_id so it's visible to non-technical user.
2. Overide do_draft method to set cancel_reason_id to False when Set to unconfirmed

andhit-r commented:
@pedrobaeza I have submit PR to your branch https://github.com/pedrobaeza/event/pull/1. I did:
1.  Move cancel_reason_id so it can be visible by non-technical feature users.
   Cancelled registration:
   ![selection_1](https://cloud.githubusercontent.com/assets/1947376/13028710/b797791a-d2a9-11e5-908e-8fb2ca7773dd.png)
   Non-cancelled registration:
   ![selection_2](https://cloud.githubusercontent.com/assets/1947376/13028712/c7e7c766-d2a9-11e5-98e7-e3c5d0f76f4e.png)
2. Overide **do_draft** method so it set **cancel_reason_id** to **False** when user click **set to unconfirmed**

pedrobaeza committed:
Merge pull request #1 from andhit-r/8.0-event_registration_cancel_reason

Move cancel_reason_id & overide do_draft method

pedrobaeza commented:
Thanks for your help, Andhitia! I have merged your PR and I think this is now ready to be merged.

dreispt commented on event_registration_cancel_reason/__openerp__.py:
```diff
@@ -0,0 +1,24 @@
+              "Serv. Tecnol. Avanzados - Pedro M. Baeza, "
+              "Odoo Community Association (OCA)",
+    "website": "https://www.antiun.com",
+    "category": "Customer Relationship Management",
```
Event?

dreispt commented on event_registration_cancel_reason/models/event_registration.py:
```diff
@@ -0,0 +1,42 @@
+
+    cancel_reason_id = fields.Many2one(
+        comodel_name='event.registration.cancel.reason', readonly=True,
+        string="Reason for cancellation", ondelete="restrict")
```
I suggest the text "Cancellation reason"

dreispt commented on event_registration_cancel_reason/models/event_registration.py:
```diff
@@ -0,0 +1,42 @@
+        }
+
+    @api.one
+    def do_draft(self):
```
This method name does not express well what it does.
What about something like `button_reg_cancel_undo`?

pedrobaeza replied:
I'm afraid this is not something we can choose, because it's the name that Odoo gave it.

dreispt replied:
OK

dreispt commented on event_registration_cancel_reason/wizard/event_registration_cancel_log_reason.py:
```diff
@@ -0,0 +1,35 @@
+
+    @api.multi
+    def button_log(self):
+        self.ensure_one()
```
It would be nice in the future to support multi selection.
Maybe you could add that idea to the roadmap section in the README.

pedrobaeza replied:
Multi-record support for wizards (TransientModel) doesn't make too much sense. I always put self.ensure_one() to avoid misuses from XML-RPC for example.

elicoidal replied:
Agree with @dreispt here: it could be good to be able to select multiple records at once. Probably could be mentioned in the roadmap

pedrobaeza replied:
@elicoidal, please read my previous comment. Wizards don't have multiple records. You instance them only once.

elicoidal replied:
@pedrobaeza I know.
this is why I suggest the roadmap to add the feature accessible in tree view. For me it makes sense to be able possibility to mass cancel registration selecting registrations in the tree view (with multi-select/Action/Cancel Registration) with similar effect as clicking on the red cross in the following view and adding a reason during the wizard.
![Uploading Screenshot from 2016-02-14 21-12-03.png…]()

pedrobaeza replied:
OK, I think we are talking about different things: the `self.ensure_one()` is there by the reason exposed and it's correct. But you're talking about reading `active_ids` variable from context instead of `active_id` to support multi-records. AFAIK, there's no this possibility now in the core, but I can enable it very easily changing the browse instruction. Do we understand now?

elicoidal replied:
Yes: no problem kept on my side :)

dreispt replied:
@pedrobaeza  Correct. AFAIK with `active_ids` a wizard can operate on many records, and adding that is possible. But we are not asking you to implement it; adding the idea to the roadmap is good enough.

pedrobaeza replied:
It's already done :wink:

dreispt commented:
Minor notes, and it LGTM :+1:

pedrobaeza committed:
[FIX] event_registration_cancel_reason: Module category/Multi support/text

pedrobaeza commented:
@dreispt, thanks for your review! Some of them have been fixed, and the others commented.

elicoidal commented on event_registration_cancel_reason/README.rst:
```diff
@@ -0,0 +1,70 @@
+   :alt: License: AGPL-3
+
+=============================================
+Reasons for event registrations cancellations
```
Reasons for event registration cancellation?

elicoidal commented on event_registration_cancel_reason/README.rst:
```diff
@@ -0,0 +1,70 @@
+Reasons for event registrations cancellations
+=============================================
+
+This module allows to add a predefined reason why a registration for an event
```
This module allows to add predefined reasons for the attendees to report back why they cancel their registration to an event.

elicoidal commented on event_registration_cancel_reason/README.rst:
```diff
@@ -0,0 +1,70 @@
+
+.. image:: http://odoo-community.org/logo.png
+   :alt: Odoo Community Association
+   :target: http://odoo-community.org
```
https (x2)

elicoidal commented on event_registration_cancel_reason/models/event_registration.py:
```diff
@@ -0,0 +1,42 @@
+
+    cancel_reason_id = fields.Many2one(
+        comodel_name='event.registration.cancel.reason', readonly=True,
+        string="Cancellation reason", ondelete="restrict")
```
Be consistent: everywhere it should be "Cancellation reasons" (instead of Cancel Reason)

pedrobaeza replied:
No, this time you are selecting only one reason, not multiples, so it's correct.

elicoidal replied:
yes: you are right here. My point is more about the "Cancel Reason" which is not so good

pedrobaeza replied:
Are you referring to the technical field name. I think this one is correct, because if not, the name should be more difficult to type in code and long without reasons. All the user labels should be with the full string though.

elicoidal replied:
I am talking indeed about the strings (which is the frontend display name).
Other names (variables etc.) needs to be clean but keeping them practical

pedrobaeza replied:
OK, I think they are all changed. Please tell me if not.

elicoidal commented on event_registration_cancel_reason/models/event_registration.py:
```diff
@@ -0,0 +1,42 @@
+    event_type_ids = fields.Many2many(
+        comodel_name="event.type", string="Event types",
+        help="Select the event types where you want to use this cancellation "
+             "reason. Leave it empty for using in all")
```
period

elicoidal commented on event_registration_cancel_reason/views/event_registration_view.xml:
```diff
@@ -0,0 +1,56 @@
+        <field name="inherit_id" ref="event.view_registration_search"/>
+        <field name="arch" type="xml">
+            <filter context="{'group_by':'state'}" position="after">
+                <filter string="Cancel reason" domain="[]" context="{'group_by': 'cancel_reason_id'}"/>
```
Cancellation reason

elicoidal commented on event_registration_cancel_reason/views/event_registration_view.xml:
```diff
@@ -0,0 +1,56 @@
+    </record>
+
+    <record model="ir.ui.view" id="event_registration_cancel_reason_tree">
+        <field name="name">Registrations cancel reasons tree</field>
```
Registration cancellation reasons tree

elicoidal commented on event_registration_cancel_reason/views/event_registration_view.xml:
```diff
@@ -0,0 +1,56 @@
+        <field name="name">Registrations cancel reasons tree</field>
+        <field name="model">event.registration.cancel.reason</field>
+        <field name="arch" type="xml">
+            <tree name="Registrations cancel reasons" editable="top">
```
Registration cancellation reasons

elicoidal commented on event_registration_cancel_reason/views/event_registration_view.xml:
```diff
@@ -0,0 +1,56 @@
+    </record>
+
+    <record id="event_registration_cancel_reason_action" model="ir.actions.act_window">
+        <field name="name">Registrations cancel reasons</field>
```
Registration cancellation reasons

elicoidal commented on event_registration_cancel_reason/wizard/event_registration_cancel_log_reason_view.xml:
```diff
@@ -0,0 +1,30 @@
+    <data>
+
+    <record model="ir.ui.view" id="event_registration_cancel_log_reason_view">
+        <field name="name">Wizard for logging registrations cancellation</field>
```
Logging registration cancellation wizard

pedrobaeza commented:
Hi, @elicoidal, I think I have attended all your comments, and two of them has been commented back. Let me know if there's anything more.

elicoidal commented on event_registration_cancel_reason/models/event_registration.py:
```diff
@@ -0,0 +1,42 @@
+            return super(EventRegistration, self).button_reg_cancel()
+        return {
+            'type': 'ir.actions.act_window',
+            'name': 'Cancel reason',
```
Cancellation reason

elicoidal commented on event_registration_cancel_reason/views/event_registration_view.xml:
```diff
@@ -0,0 +1,56 @@
+    <data>
+
+    <record model="ir.ui.view" id="view_event_registration_form">
+        <field name="name">Event registration form (with cancel reason)</field>
```
Cancellation

elicoidal commented on event_registration_cancel_reason/views/event_registration_view.xml:
```diff
@@ -0,0 +1,56 @@
+            <data>
+                <xpath expr="//sheet/h1" position="after">
+                    <h2 attrs="{'invisible':[('cancel_reason_id','=',False)]}">
+                        <label for="cancel_reason_id" string="Cancel Reason:"/>
```
Cancellation

elicoidal commented on event_registration_cancel_reason/views/event_registration_view.xml:
```diff
@@ -0,0 +1,56 @@
+    </record>
+
+    <record model="ir.ui.view" id="view_registration_search">
+        <field name="name">Event registration search (with cancel reason)</field>
```
Cancellation

elicoidal commented on event_registration_cancel_reason/README.rst:
```diff
@@ -0,0 +1,70 @@
+Configuration
+=============
+
+Go to *Marketing > Configuration > Events > Registrations cancel reasons* for
```
Cancellation reasons

elicoidal commented on event_registration_cancel_reason/README.rst:
```diff
@@ -0,0 +1,70 @@
+registration is cancelled.
+
+You will get also the same screen when pressing the button
+*Cancel registration* from the details view of a registration in
```
Cancellation

elicoidal commented on event_registration_cancel_reason/README.rst:
```diff
@@ -0,0 +1,70 @@
+You will get also the same screen when pressing the button
+*Cancel registration* from the details view of a registration in
+*Marketing > Events > Registrations*. On this object, you can group by the
+cancel reason in the list, calendar or graph view.
```
Cancellation

andhit-r committed:
String change: cancel -> cancellation

pedrobaeza committed:
Merge pull request #2 from andhit-r/8.0-event_registration_cancel_reason

String change: cancel -> cancellation

pedrobaeza commented:
Thanks to @andhit-r now texts should be fixed.

yajo committed:
Allow multirecord cancelling and fix tests.

yajo commented:
I fixed travis for you, please merge.

yajo committed:
Provide a better error message and test it.

yajo committed:
Go LBYL.

pedrobaeza committed:
Merge pull request #4 from Yajo/PR-30

Allow multirecord cancelling and fix tests.

pedrobaeza commented:
I think it should be all right now.

max3903 commented:
:+1:

pedrobaeza commented:
This one can be merged.
